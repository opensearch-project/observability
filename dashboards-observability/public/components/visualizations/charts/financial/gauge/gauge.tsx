/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { find, isEmpty } from 'lodash';
import Plotly from 'plotly.js-dist';
import React, { useMemo } from 'react';
import {
  AGGREGATIONS,
  GROUPBY,
  PLOTLY_GAUGE_COLUMN_NUMBER,
  DEFAULT_GAUGE_CHART_PARAMETERS,
} from '../../../../../../common/constants/explorer';
import { IVisualizationContainerProps } from '../../../../../../common/types/explorer';
import { PLOT_MARGIN } from '../../../../../../common/constants/shared';
import { ThresholdUnitType } from '../../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds';
import { EmptyPlaceholder } from '../../../../event_analytics/explorer/visualizations/shared_components/empty_placeholder';
import { getPropName } from '../../../../event_analytics/utils/utils';
import { Plt } from '../../../plotly/plot';

const {
  GaugeTitleSize,
  DisplayDefaultGauges,
  OrientationDefault,
  TickLength,
  LegendPlacement,
} = DEFAULT_GAUGE_CHART_PARAMETERS;

export const Gauge = ({ visualizations, layout, config }: any) => {
  const {
    data: {
      rawVizData: {
        data: queriedVizData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          span = {},
          chartStyles = {},
          panelOptions = {},
          thresholds = [],
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        },
        layoutConfig = {},
      },
    },
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const seriesLength = series.length;
  const numberOfGauges = panelOptions.numberOfGauges || DisplayDefaultGauges;

  // style parameters
  const titleSize = chartStyles.titleSize || GaugeTitleSize;
  const valueSize = chartStyles.valueSize;
  const showThresholdMarkers = chartStyles.showThresholdMarkers || false;
  const showThresholdLabels = chartStyles.showThresholdLabels || false;
  const orientation = chartStyles.orientation || OrientationDefault;
  const legendPlacement = chartStyles.legendPlacement || LegendPlacement;

  let xaxes = dimensions;
  const isEmptyPlot = !seriesLength || isEmpty(queriedVizData);
  const timestampField = find(fields, (field) => field.type === 'timestamp');
  if (span && span.time_field && timestampField) {
    xaxes = [timestampField, ...xaxes];
  }

  if (isEmptyPlot) return <EmptyPlaceholder icon={icontype} />;

  const gaugeData: Plotly.Data[] = useMemo(() => {
    let calculatedGaugeData: Plotly.Data[] = [];
    // case 1,2: no dimension, single/multiple metrics
    if (!xaxes.length && seriesLength >= 1) {
      calculatedGaugeData = series.map((seriesItem: any) => {
        return {
          field_name: getPropName(seriesItem),
          value: queriedVizData[getPropName(seriesItem)][0],
        };
      });
    }

    // case 3: multiple dimensions and multiple metrics
    if (xaxes.length && seriesLength) {
      const selectedDimensionsData = xaxes
        .map((dimension: any) => {
          return queriedVizData[dimension.name]
            ? queriedVizData[dimension.name].slice(0, numberOfGauges)
            : [];
        })
        .reduce((prev, cur) => {
          return prev.map((i, j) => `${i}, ${cur[j]}`);
        });
      const selectedSeriesData = series.map((seriesItem: any) => {
        const propValue = getPropName(seriesItem);
        return queriedVizData[`${propValue}`]
          ? queriedVizData[`${propValue}`].slice(0, numberOfGauges)
          : [];
      });

      selectedSeriesData.map((seriesSlice: any, seriesSliceIndex: number) => {
        calculatedGaugeData = [
          ...calculatedGaugeData,
          ...seriesSlice.map((seriesSliceData: any, seriesSliceDataIndex: number) => {
            return {
              field_name: `${selectedDimensionsData[seriesSliceDataIndex]}, ${getPropName(
                series[seriesSliceIndex]
              )}`,
              value: seriesSliceData,
            };
          }),
        ];
      });
    }

    return calculatedGaugeData.map((gauge, index) => {
      return {
        type: 'indicator',
        mode: 'gauge+number+delta',
        value: gauge.value || 0,
        title: {
          text: gauge.field_name,
          font: { size: titleSize },
          align: legendPlacement,
        },
        ...(valueSize && {
          number: {
            font: {
              size: valueSize,
            },
          },
        }),
        domain: {
          ...(orientation === 'auto' || orientation === 'h'
            ? {
                row: Math.floor(index / PLOTLY_GAUGE_COLUMN_NUMBER),
                column: index % PLOTLY_GAUGE_COLUMN_NUMBER,
              }
            : {
                column: Math.floor(index / PLOTLY_GAUGE_COLUMN_NUMBER),
                row: index % PLOTLY_GAUGE_COLUMN_NUMBER,
              }),
        },
        gauge: {
          ...(showThresholdMarkers &&
            thresholds &&
            thresholds.length && {
              threshold: {
                line: { color: thresholds[0]?.color || 'red', width: 4 },
                thickness: 0.75,
                value: thresholds[0]?.value || 0,
              },
            }),
          // threshold labels
          ...(showThresholdLabels && thresholds && thresholds.length
            ? {
                axis: {
                  ticktext: [gauge.value, ...thresholds.map((t: ThresholdUnitType) => t.name)],
                  tickvals: [gauge.value, ...thresholds.map((t: ThresholdUnitType) => t.value)],
                  ticklen: TickLength,
                },
              }
            : {}),
        },
      };
    });
  }, [
    xaxes,
    series,
    queriedVizData,
    fields,
    thresholds,
    showThresholdMarkers,
    orientation,
    showThresholdLabels,
    titleSize,
    valueSize,
  ]);

  const mergedLayout = useMemo(() => {
    const isAtleastOneFullRow = Math.floor(gaugeData.length / PLOTLY_GAUGE_COLUMN_NUMBER) > 0;
    return {
      grid: {
        ...(orientation === 'auto' || orientation === 'h'
          ? {
              rows: Math.floor(gaugeData.length / PLOTLY_GAUGE_COLUMN_NUMBER) + 1,
              columns: isAtleastOneFullRow ? PLOTLY_GAUGE_COLUMN_NUMBER : gaugeData.length,
            }
          : {
              columns: Math.floor(gaugeData.length / PLOTLY_GAUGE_COLUMN_NUMBER) + 1,
              rows: isAtleastOneFullRow ? PLOTLY_GAUGE_COLUMN_NUMBER : gaugeData.length,
            }),
        pattern: 'independent',
      },
      ...layout,
      ...(layoutConfig.layout && layoutConfig.layout),
      title: panelOptions.title || layoutConfig.layout?.title || '',
      margin: {
        ...PLOT_MARGIN,
        t: 100,
      },
    };
  }, [layout, gaugeData.length, layoutConfig.layout, panelOptions.title, orientation]);

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };
  return <Plt data={gaugeData} layout={mergedLayout} config={mergedConfigs} />;
};
