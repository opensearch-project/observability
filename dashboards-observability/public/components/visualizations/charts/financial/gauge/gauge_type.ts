/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Gauge } from './gauge';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../../shared/shared_configs';
import { LensIconChartLine } from '../../../assets/chart_line';
import { VizDataPanel } from '../../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigThresholds,
  InputFieldItem,
  SwitchButton,
  ConfigChartOptions,
  ButtonGroupItem,
} from '../../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import { DEFAULT_GAUGE_CHART_PARAMETERS } from '../../../../../../common/constants/explorer';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();
const { ThresholdsMaxLimit } = DEFAULT_GAUGE_CHART_PARAMETERS;

export const createGaugeTypeDefinition = (params: any = {}) => ({
  name: 'Gauge',
  type: 'indicator',
  id: 'gauge',
  label: 'Gauge',
  fulllabel: 'Gauge',
  icontype: 'visGauge',
  category: VIS_CATEGORY.BASICS,
  selection: {
    dataLoss: 'nothing',
  },
  icon: LensIconChartLine,
  valueseries: 'yaxis',
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          {
            id: 'chart-styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Orientation',
                component: ButtonGroupItem,
                mapTo: 'orientation',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Auto', id: 'auto' },
                    { name: 'Vertical', id: 'v' },
                    { name: 'Horizontal', id: 'h' },
                  ],
                  defaultSelections: [{ name: 'Auto', id: 'auto' }],
                },
              },
              {
                name: 'Legend placement',
                component: ButtonGroupItem,
                mapTo: 'legendPlacement',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Center', id: 'center' },
                    { name: 'Right', id: 'right' },
                    { name: 'Left', id: 'left' },
                  ],
                  defaultSelections: [{ name: 'Center', id: 'center' }],
                },
              },
              {
                title: 'Title size',
                name: 'Title size',
                component: InputFieldItem,
                mapTo: 'titleSize',
                eleType: 'input',
              },
              {
                title: 'Value size',
                name: 'Value size',
                component: InputFieldItem,
                mapTo: 'valueSize',
                eleType: 'input',
              },
              {
                title: 'Show threshold labels',
                name: 'Show threshold labels',
                component: SwitchButton,
                mapTo: 'showThresholdLabels',
                eleType: 'switchButton',
                currentValue: false,
              },
              {
                title: 'Show threshold markers',
                name: 'Show threshold markers',
                component: SwitchButton,
                mapTo: 'showThresholdMarkers',
                eleType: 'switchButton',
                currentValue: true,
              },
            ],
          },
          {
            id: 'thresholds',
            name: 'Thresholds',
            editor: ConfigThresholds,
            mapTo: 'thresholds',
            defaultState: [],
            schemas: [],
            props: {
              maxLimit: ThresholdsMaxLimit,
            },
          },
        ],
      },
      {
        id: 'style-panel',
        name: 'Layout',
        mapTo: 'layoutConfig',
        editor: ConfigEditor,
        content: [],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
  },
  component: Gauge,
});
