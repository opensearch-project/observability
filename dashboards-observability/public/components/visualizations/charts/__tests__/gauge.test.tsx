/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Gauge } from '../financial/gauge/gauge';
import {
  LAYOUT_CONFIG,
  GAUGE_TEST_VISUALIZATIONS_DATA,
} from '../../../../../test/event_analytics_constants';

describe('Gauge component', () => {
  configure({ adapter: new Adapter() });

  it('Renders gauge component', async () => {
    const wrapper = mount(
      <Gauge visualizations={GAUGE_TEST_VISUALIZATIONS_DATA} layout={LAYOUT_CONFIG} config={true} />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
