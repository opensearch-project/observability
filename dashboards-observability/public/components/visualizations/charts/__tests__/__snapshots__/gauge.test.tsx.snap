// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gauge component Renders gauge component 1`] = `
<Gauge
  config={true}
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {
          "dataConfig": Object {
            "valueOptions": Object {
              "dimensions": Array [
                Object {
                  "label": "tags",
                  "name": "tags",
                  "type": "text",
                },
              ],
              "metrics": Array [
                Object {
                  "label": "count()",
                  "name": "count()",
                  "side": "left",
                  "type": "integer",
                },
              ],
            },
          },
        },
      },
      "vis": Object {
        "category": "Visualizations",
        "component": [Function],
        "editorconfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Style",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "chart-styles",
                  "mapTo": "chartStyles",
                  "name": "Chart styles",
                  "schemas": Array [
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                          Object {
                            "id": "v",
                            "name": "Vertical",
                          },
                          Object {
                            "id": "h",
                            "name": "Horizontal",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "legendPlacement",
                      "name": "Legend placement",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "center",
                            "name": "Center",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "center",
                            "name": "Center",
                          },
                          Object {
                            "id": "right",
                            "name": "Right",
                          },
                          Object {
                            "id": "left",
                            "name": "Left",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "titleSize",
                      "name": "Title size",
                      "title": "Title size",
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "valueSize",
                      "name": "Value size",
                      "title": "Value size",
                    },
                    Object {
                      "component": [Function],
                      "currentValue": false,
                      "eleType": "switchButton",
                      "mapTo": "showThresholdLabels",
                      "name": "Show threshold labels",
                      "title": "Show threshold labels",
                    },
                    Object {
                      "component": [Function],
                      "currentValue": true,
                      "eleType": "switchButton",
                      "mapTo": "showThresholdMarkers",
                      "name": "Show threshold markers",
                      "title": "Show threshold markers",
                    },
                  ],
                },
                Object {
                  "defaultState": Array [],
                  "editor": [Function],
                  "id": "thresholds",
                  "mapTo": "thresholds",
                  "name": "Thresholds",
                  "props": Object {
                    "maxLimit": 1,
                  },
                  "schemas": Array [],
                },
              ],
            },
            Object {
              "content": Array [],
              "editor": [Function],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
          ],
        },
        "fulllabel": "Gauge",
        "icon": [Function],
        "icontype": "visGauge",
        "id": "gauge",
        "label": "Gauge",
        "name": "Gauge",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "type": "indicator",
        "valueseries": "yaxis",
        "visconfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "layout": Object {
            "height": 1180,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <EmptyPlaceholder
    icon="visGauge"
  >
    <EuiText
      className="lnsChart__empty"
      color="subdued"
      data-test-subj="vizWorkspace__noData"
      size="xs"
      textAlign="center"
    >
      <div
        className="euiText euiText--extraSmall lnsChart__empty"
        data-test-subj="vizWorkspace__noData"
      >
        <EuiTextAlign
          textAlign="center"
        >
          <div
            className="euiTextAlign euiTextAlign--center"
          >
            <EuiTextColor
              color="subdued"
              component="div"
            >
              <div
                className="euiTextColor euiTextColor--subdued"
              >
                <EuiIcon
                  color="subdued"
                  size="xxl"
                  type="visGauge"
                >
                  <EuiIconEmpty
                    aria-hidden={true}
                    className="euiIcon euiIcon--xxLarge euiIcon--subdued euiIcon-isLoading"
                    focusable="false"
                    role="img"
                    style={null}
                  >
                    <svg
                      aria-hidden={true}
                      className="euiIcon euiIcon--xxLarge euiIcon--subdued euiIcon-isLoading"
                      focusable="false"
                      height={16}
                      role="img"
                      style={null}
                      viewBox="0 0 16 16"
                      width={16}
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </EuiIconEmpty>
                </EuiIcon>
                <EuiSpacer
                  size="l"
                >
                  <div
                    className="euiSpacer euiSpacer--l"
                  />
                </EuiSpacer>
                <p>
                  <FormattedMessage
                    defaultMessage="No data found"
                    id="visualization_noData"
                    values={Object {}}
                  >
                    <span>
                      No data found
                    </span>
                  </FormattedMessage>
                </p>
              </div>
            </EuiTextColor>
          </div>
        </EuiTextAlign>
      </div>
    </EuiText>
  </EmptyPlaceholder>
</Gauge>
`;
